core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java

```diff
@@ -0,0 +1,344 @@
+            Types.NestedField.optional(2, "long", Types.DoubleType.get())))
+        .commit();
+
+    String updatedJson = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    NameMapping updated = NameMappingParser.fromJson(updatedJson);
+
+    validateUnchanged(mapping, updated);
+
+    MappedField newMapping = updated.find("location");
+    Assert.assertNotNull("Mapping for new column should be added", newMapping);
+
+    Assert.assertEquals("Mapping should use the assigned field ID",
+        (Integer) table.schema().findField("location").fieldId(), updated.find("location").id());
+    Assert.assertNotNull("Should contain a nested mapping", updated.find("location").nestedMapping());
+
+    Assert.assertEquals("Mapping should use the assigned field ID",
+        (Integer) table.schema().findField("location.lat").fieldId(), updated.find("location.lat").id());
+    Assert.assertNull("Should not contain a nested mapping", updated.find("location.lat").nestedMapping());
+
+    Assert.assertEquals("Mapping should use the assigned field ID",
+        (Integer) table.schema().findField("location.long").fieldId(), updated.find("location.long").id());
+    Assert.assertNull("Should not contain a nested mapping", updated.find("location.long").nestedMapping());
+  }
+
+  @Test
+  public void testRenameColumn() {
+    NameMapping mapping = MappingUtil.create(table.schema());
+    String mappingJson = NameMappingParser.toJson(mapping);
+
+    table.updateProperties()
+        .set(TableProperties.DEFAULT_NAME_MAPPING, mappingJson)
+        .commit();
+
+    table.updateSchema()
+        .renameColumn("id", "object_id")
+        .commit();
+
+    String updatedJson = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    NameMapping updated = NameMappingParser.fromJson(updatedJson);
+
+    int idColumnId = table.schema().findField("object_id").fieldId();
+    validateUnchanged(
+        Iterables.filter(mapping.asMappedFields().fields(), field -> !Objects.equals(idColumnId, field.id())),
+        updated);
+
+    MappedField updatedMapping = updated.find(idColumnId);
+    Assert.assertNotNull("Mapping for id column should exist", updatedMapping);
+    Assert.assertEquals("Should add the new column name to the existing mapping",
+        MappedField.of(idColumnId, ImmutableList.of("id", "object_id")),
+        updatedMapping);
+  }
+
+  @Test
+  public void testDeleteColumn() {
+    NameMapping mapping = MappingUtil.create(table.schema());
+    String mappingJson = NameMappingParser.toJson(mapping);
+
+    table.updateProperties()
+        .set(TableProperties.DEFAULT_NAME_MAPPING, mappingJson)
+        .commit();
+
+    table.updateSchema()
+        .deleteColumn("id")
+        .commit();
+
+    String updatedJson = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
+    NameMapping updated = NameMappingParser.fromJson(updatedJson);
+
+    // should not change the mapping
```

Comment:
Why we don't want to remove deleted entry in mapping?

Reply:
Deleting the entry would remove information that may be valuable: the association of a now-deleted ID with a set of names. That can be used in some mapping cases.

For example, we use mappings from JSON records in Kafka into Iceberg data. If a user drops a column, the incoming data is dropped as well. Using the mapping to drop data gives us more flexibility in how to handle un-mapped data. We move un-mapped fields to a string-map called `other_properties`, but you could also detect that there is un-mapped data coming in and update a metric or decide to pause until the mapping is updated.